### PR TITLE
devito to xdsl: first draft of extending Devito IR to XDSL 

### PR DESCRIPTION
**Pull request by stoltzstrop**
The work in this PR extends what was started in [this jupyter notebook](https://github.com/xdslproject/devito/blob/6629f6c25733afefa5f90860cbbe0397e9c6d4e9/GenerateXDSL.ipynb) to convert a simple devito example to xdsl.

The updated example can be found in: `devito_to_xdsl_examples/simple_example_ietxdsl.py`

Some of the main remaining issues include the following: 

- The example does not produce code directly from `op`, but instead cherry-picks the headers, includes, struct declarations and then the computing kernel itself from this variable
-- In addition to this, there seem to be some inconsistencies in both the variable names required and their types (ie. `u` does not have a straightforward type to pull out like the other kernel parameter values and `t0` and `t1` must also be passed in initially - to get around the latter issue I've added extra fields to `devito/ir/ietxdsl/operations::Callable`, but this is a bit awkward)
- Some new IR additions (ie. `PointerCast`) are a bit light on detail and basically just pass the C code over
- Converting `node.Section`s seems to be a bit awkward

There are almost certainly many other improvements to be made here, but am just opening this up mainly as a talking point for further progression

### CODE REVIEW

File: devito_to_xdsl_examples/simple_example_ietxdsl.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+# those parameters without associated types aren't printed in the Kernel header
+op_params = list(op.parameters)
+op_param_names = [opi.name for opi in op_params]
+op_header_params = [opi.name for opi in op_params]
+op_types = [opi._C_typedata for opi in op_params]
+
+# TODO: u also does not get the right type and must be passed in this way for now
+op_types[0] = str(op_params[0]._C_typename)
+
+# TODO: we still need to add "t0" and "t1" even though they aren't passed in
```
[the remark this answers is not in the document]
Comment (stoltzstrop) [reply devito_to_xdsl_examples/simple_example_ietxdsl.py:38]: I'm not sure how else to make the types available to the kernel function header..do you know of a better option?

File: devito/ir/ietxdsl/cgeneration.py [hunk 2]
```diff
@@ -31,15 +54,199 @@ def printModule(self, module):
...
+        self.dedent()
+        self.print("}")
+        pass
+
+    def printIterationWithSubIndices(self,
+                                     iteration_op: IterationWithSubIndices):
+        uindices_names = iteration_op.uindices_names
+        uindices_symbmins = iteration_op.uindices_symbmins
+        ssa_val = iteration_op.body.blocks[0].args[0]
+        iterator = "i_" + str(len(self.iterator_names))
```
Comment (georgebisbas) [inline devito/ir/ietxdsl/cgeneration.py:109]: iterator = str(iteration_op.arg_name.data) ?

File: .github/workflows/pytest-core-nompi.yml [hunk 3]
```diff
@@ -80,4 +80,4 @@ jobs:
     - name: Test with pytest
       run: |
         ${{ steps.set-run.outputs.RUN_CMD }} ${{ matrix.arch }} --version
-        ${{ steps.set-run.outputs.RUN_CMD }} pytest -k "${{ matrix.test-set }}" -m "not parallel" --cov --cov-config=.coveragerc --cov-report=xml ${{ steps.set-tests.outputs.TESTS }}
\ No newline at end of file
+        ${{ steps.set-run.outputs.RUN_CMD }} pytest -k "${{ matrix.test-set }}" -m "not parallel" --cov --cov-config=.coveragerc --cov-report=xml ${{ steps.set-tests.outputs.TESTS }}
```
Comment (georgebisbas) [inline .github/workflows/pytest-core-nompi.yml:83]: we need this empty line back, just to avoid this not needed diff

**Review by georgebisbas — COMMENTED** [review]
Thanks, Larissa!
I think this gets close to being merged into the xdsl-devito fork.
my next step should be to form this 
xdsl_string = transform_devito_xdsl_string(op)
as a pass in the compiler.

File: tests/test_iet_xdsl_examples.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+from devito.ir.ietxdsl.xdsl_passes import transform_devito_xdsl_string
+# flake8: noqa
+
+def test_udx_conversion():
+
+    # Define a simple Devito Operator
+    grid = Grid(shape=(3, 3))
+    u = TimeFunction(name='u', grid=grid)
+    eq = Eq(u.forward, u.dx)
+    op = Operator([eq])
```
Comment (georgebisbas) [inline tests/test_iet_xdsl_examples.py:11]: TBD: 
op = Operator ([eq], opt = ("xdsl"))

**Pull request merged**